deps/chakrashim/src/v8handlescope.cc

```diff
@@ -54,28 +54,38 @@ HandleScope *HandleScope::GetCurrent() {
 }
 
 bool HandleScope::AddLocal(JsValueRef value) {
-  if (_count < _countof(_locals)) {
+  if (_count <= kOnStackLocals) {
     _locals[_count++] = value;
-    return true;
+    if (_count < kOnStackLocals) {
+      return true;
+    }
   }
 
+  // Ran out of _locals. Push everything in array and empty
+  // the _locals array.
   if (_refs == JS_INVALID_REFERENCE) {
     if (JsCreateArray(1, &_refs) != JsNoError) {
       return AddLocalAddRef(value);
     }
   }
 
-  JsValueRef index;
+  JsValueRef saveInHandleScopeFunction =
+    jsrt::ContextShim::GetCurrent()->GetsaveInHandleScopeFunction();
 
-  if (JsIntToNumber(_count - _countof(_locals), &index) != JsNoError) {
-    return AddLocalAddRef(value);
+  JsValueRef args[kOnArgsToPush];
+  args[0] = jsrt::GetUndefined();
+  args[1] = _refs;
+  for (int index = 0; index < _count;index++) {
+    args[index + 2] = _locals[index];
+    _locals[index] = JS_INVALID_REFERENCE;
   }
-
-  if (JsSetIndexedProperty(_refs, index, value) != JsNoError) {
+  JsValueRef result;
+  if (JsCallFunction(saveInHandleScopeFunction, args,
```

Comment:
If this call fails it'll be in trouble as you already cleared _locals cache. Looks you can only clear it after this succeeds.